rfcs/xxxx-collage.md

```diff
@@ -0,0 +1,863 @@
+                        v                   v                       v
+                                      UnionFusionRule
+                                            |
+                                            v
+                             CombineByPrimitivesFusionRule (with default TVM primitive rules)
+```
+
+### FusionSpec
+
+A `FusionSpec` pairs a a `FusionRule` with a `Target`.
+
+### Phase 1
+
+We build on the existing TVM support for heterogeneous devices and targets. The available `Targets` are extracted from
+the compilation configuration (eg using the existing `CompilationConfig` helper class). Each target is inspected to
+decide on how to construct a `FusionSpec`, which will guide Collage in the selection of candidate kernels to explore for
+that target.
+
+- If the `Target` has a `"fusion_spec"` attribute, use that directly (not currently in the 'v2' prototype). This would
+  allow users to directly control fusion for the target's they care about.
+- If the `Target` has a `"compiler"` attribute (eg `"cutlass"`), and the global pattern table has an entry for that
+  attribute value, assume the `Target` denotes a pattern-based BYOC integration to explore. The `FusionSpec`
+  will import all the BYOC patterns and predicates automatically.
+- As above, but if global pattern has no matching entry, assume the `Target` denotes a predicate-based BYOC integration
+  to explore (eg `"tensorrt"`). The `FusionSpec` will look for and evaluate predicates with the
+  `"target.<compiler>"` attribute on all Relay operators.
+- Otherwise, assume the `Target` denotes a TVM-native target. The `FusionSpec` mimics the existing `FuseOps`, but now
+  generalized to explore multiple candidates so as to leave room for possible BYOC candidates.
+
+Note that to make this approach work we need to allow for multiple `Target`s with the same `DLDeviceKind`. For the VM
+simply switching the `target` argument from dictionary to list and removing some redundant Python preprocessing code was
+all that was required to support this.
+
+The user can use `on_device` annotations to constrain sub-graphs to particular devices. When Collage is considering
+candidate kernels, it should be sure to choose a candidate `Target` which 'refines' the `Target` for every
+sub-expression discovered by the `PlanDevicesPass`. Given targets T and U we say 'T refines U' if T has a
+'"compiler"' and/or '"fusion_spec"' attributes, U has no such attributes, and T and U otherwise agree on all other
+fields. (This is not currently in the 'v2' prototype).
+
+### Phase 2
+
+Most of the hard work for this phase is carried by the `AllCandidateKernels` implementations of the `FusionRule`s. The
+main driver simply needs to index all the found `CandidateKernels` by their minimum 'inside' `PostDfsIndex`
+for rapid retrieval during the shortest path search.
+
+### Phase 3
+
+We find it most natural to use Dijkstra to find the optimal partitioning. A `SearchState` is:
```

Comment:
I think the cascader has a similar algorithm as part of the 'proposal generator', but not using Dijkstra's (just a top-down dynamic programming solution). It'd be interesting to work out if that'd be applicable here, and if so whether it's faster or not.

Reply:
My hope was the search bloom with Dijkstra would save us from measuring candidates which ultimately wont be in the optimal partitioning. However we not yet ready to measure the bloom on actual largish models.

I spent the last few years working on route optimization for worldwide transport networks so that particular hammer is most familiar to me :-)

Reply:
Going to resolve but we can adjust search algorithm  once we come up against the reality of search on large models.